doc/source/development/rfc/rfc96_deferred_plugin_loading.rst

```diff
@@ -0,0 +1,449 @@
+     * <li>GDAL_DMD_EXTENSION</li>
+     * <li>GDAL_DCAP_RASTER</li>
+     * <li>GDAL_DCAP_MULTIDIM_RASTER</li>
+     * <li>GDAL_DCAP_VECTOR</li>
+     * <li>GDAL_DCAP_GNM</li>
+     * <li>GDAL_DMD_OPENOPTIONLIST</li>
+     * <li>GDAL_DMD_SUBDATASETS</li>
+     * <li>GDAL_DCAP_MULTIPLE_VECTOR_LAYERS</li>
+     * <li>GDAL_DCAP_NONSPATIAL</li>
+     * </ul>
+     *
+     * The pfnIdentify and pfnGetSubdatasetInfoFunc callbacks, if they are
+     * defined in the real driver, should also be set on the proxy driver.
+     *
+     * Furthermore, the following metadata items must be defined if the real
+     * driver sets the corresponding callback:
+     * <ul>
+     * <li>GDAL_DCAP_OPEN: must be set to YES if the real driver defines pfnOpen</li>
+     * <li>GDAL_DCAP_CREATE: must be set to YES if the real driver defines pfnCreate</li>
+     * <li>GDAL_DCAP_CREATE_MULTIDIMENSIONAL: must be set to YES if the real driver defines pfnCreateMultiDimensional</li>
+     * <li>GDAL_DCAP_CREATECOPY: must be set to YES if the real driver defines pfnCreateCopy</li>
+     * </ul>
+     *
+     * @since 3.9
+     */
+    class GDALPluginDriverProxy : public GDALDriver
+    {
+      public:
+        GDALPluginDriverProxy(const std::string &osPluginFileName);
+    }
+
+
+The proxy driver uses the metadata items that have been set on it
+to declare a minimum set of capabilities (GDAL_DCAP_RASTER, GDAL_DCAP_MULTIDIM_RASTER,
+GDAL_DCAP_VECTOR, GDAL_DCAP_OPEN, etc.) to which it can answer directly, and
+which are the ones used by GDALOpen() to open a dataset. For other metadata items,
+it will fallback to loading the actual driver and forward the requests to it.
+
+
+.. code-block:: cpp
+
+    /** Declare a driver that will be loaded as a plugin, when actually needed.
+     *
+     * @param poProxyDriver Plugin driver proxy
+     *
+     * @since 3.9
+     */
+     void GDALDriverManager::DeclareDeferredPluginDriver(GDALPluginDriverProxy *poProxyDriver);
+
+
+DeclareDeferredPluginDriver() method will also keep track of the plugin filename to avoid automatically
+loading it in the GDALDriverManager::AutoLoadDrivers() method (that method
+will only load out-of-tree drivers or in-tree drivers that have not been
+converted to use DeclareDeferredPluginDriver()).
+
+The main point is that drivers set the Identify() method on the proxy driver.
+That Identify() method must be compiled in libgdal itself, and thus be
+defined in a C++ file that does not depend on any external library.
+Similarly for the GetSubdatasetInfoFunc() optional method.
+
+When loading the actual driver, the GDALPluginDriverProxy::GetRealDriver()
+method will check that all information set in its metadata is
+consistent with the actual metadata of the underlying driver, and will warn
+when there are differences.
+
+GDALDataset::Open(), Create(), CreateCopy() methods are modified to not use
+directly the pfnOpen, pfnCreate, pfnCreateCopy callbacks (that would be the ones
```

Comment:
Is this change really necessary? Wouldn't be possible to delegate the loading to the pfn* in the proxy, maybe a generic function of the proxy that will load (once) the real driver and then call the real driver's pfn* ?

Reply:
I acknowledge that this is not super pretty, but it is intended to be GDAL core low-level internal detail. The issue is for example that GDALDriver::Create() currently calls pfnCreate on the proxy driver, but it doesn't know that it is a proxy driver. When the proxy driver has not loaded yet the real driver, the pfnCreate callback is null... Hence this GetCreateCallback() virtual method implemented in the proxy driver that takes care of loading the real driver and setting the pfnCreate callback of the proxy with the value of the pfnCreate of the real driver.  The alternative would be that each place that accesses the callbacks checks if the driver object derives from GDALPluginDriverProxy to force it to load the real driver, but I don't find that solution prettier.